File: frontend/src/lib/api/catalog.ts
import { apiClient } from "@/lib/api/client";

// The backend proxies these through as raw JSON from FlowPOS — types are
// best-effort, not committed field-by-field server-side.
export interface CatalogProduct {
  id: number;
  name: string;
  slug?: string;
  price: number;
  // Confirmed against a real response — there's no flat "image" field, just
  // this attachments array (empty when the product has no photo).
  attachments?: { url: string }[];
  has_variants?: boolean;
  variants_count?: number;
  default_variant_id?: number | null;
  // See utils/tax.ts for how these combine with a picked unit price.
  vat_rate?: number | null;
  is_taxable?: boolean;
  is_vat_inclusive?: boolean;
  [key: string]: unknown;
}

export function productImageUrl(product: CatalogProduct): string | undefined {
  return product.attachments?.[0]?.url;
}

export interface CatalogAddon {
  id: number;
  addon_group_id: number;
  name: string;
  price: number;
  is_active: boolean;
  max_quantity: number | null;
}

export interface CatalogAddonGroup {
  id: number;
  name: string;
  min_selection: number;
  max_selection: number;
  addons: CatalogAddon[];
}

// A variant has no flat "name" — confirmed against a real response. Its
// display name is built from `items`, one per choice dimension (e.g. a Size
// of "S" and a Color of "Red" would be two entries here).
export interface CatalogVariantItem {
  id: number;
  name: string;
  choice_type?: { id: number; label: string };
}

export interface CatalogVariant {
  id: number;
  price: number;
  sku?: string;
  items?: CatalogVariantItem[];
  add_on_groups?: CatalogAddonGroup[];
  [key: string]: unknown;
}

// Joins every choice dimension into one label ("S", or "S • Red" for a
// product with more than one choice type) — same separator convention
// tenant-dashboard's own order flow uses for multi-dimension variants.
export function variantLabel(variant: CatalogVariant): string {
  const names = (variant.items ?? []).map((item) => item.name).filter(Boolean);
  return names.length > 0 ? names.join(" • ") : `Variant #${variant.id}`;
}

// GET /catalog/products/:slug — a simple product resolves default_variant
// directly; a has_variants product lists variants to pick from.
export interface CatalogProductDetail extends CatalogProduct {
  variants?: CatalogVariant[];
  default_variant?: CatalogVariant | null;
}

export interface CatalogCategory {
  id: number;
  name: string;
  [key: string]: unknown;
}

export interface CatalogCustomer {
  id: number;
  name: string;
  email: string;
  phone?: string | null;
  [key: string]: unknown;
}

export interface CreateCustomerInput {
  name: string;
  email: string;
  phone?: string | null;
}

export interface CatalogLocation {
  id: number;
  name: string;
  type?: string;
  address?: {
    address_line_1: string;
    address_line_2?: string | null;
    city: string;
    state?: string;
    post_code: string;
    country?: { shortcode: string; name: string };
  };
  [key: string]: unknown;
}

// FlowPOS list endpoints come back as either a Laravel paginator
// ({data: T[], ...}) or (locations) a plain array — normalize both to T[].
function toArray<T>(value: unknown): T[] {
  if (Array.isArray(value)) return value as T[];
  if (value && typeof value === "object" && Array.isArray((value as Record<string, unknown>).data)) {
    return (value as { data: T[] }).data;
  }
  return [];
}

export async function listProducts(search?: string): Promise<CatalogProduct[]> {
  const { data } = await apiClient.get<{ products: unknown }>("/catalog/products", {
    params: { search },
  });
  return toArray<CatalogProduct>(data.products);
}

export async function getProduct(slug: string): Promise<CatalogProductDetail> {
  const { data } = await apiClient.get<{ product: CatalogProductDetail }>(
    `/catalog/products/${encodeURIComponent(slug)}`
  );
  return data.product;
}

export async function listCategories(search?: string): Promise<CatalogCategory[]> {
  const { data } = await apiClient.get<{ categories: unknown }>("/catalog/categories", {
    params: { search },
  });
  return toArray<CatalogCategory>(data.categories);
}

export async function listCustomers(search?: string): Promise<CatalogCustomer[]> {
  const { data } = await apiClient.get<{ customers: unknown }>("/catalog/customers", {
    params: { search },
  });
  return toArray<CatalogCustomer>(data.customers);
}

export async function createCustomer(input: CreateCustomerInput): Promise<CatalogCustomer> {
  const { data } = await apiClient.post<{ customer: CatalogCustomer }>("/catalog/customers", input);
  return data.customer;
}

export async function listLocations(search?: string): Promise<CatalogLocation[]> {
  const { data } = await apiClient.get<{ locations: unknown }>("/catalog/locations", {
    params: { search },
  });
  return toArray<CatalogLocation>(data.locations);
}

File: backend/internal/handler/catalog.go
package handler

import (
	"net/http"

	"github.com/gin-gonic/gin"

	"github.com/FlowPosLtd/quotes/backend/internal/service"
)

// CatalogHandler serves the authenticated /api/v1/catalog/* routes that
// proxy FlowPOS's own product/category/customer/location endpoints using
// the tenant's installation api_key — backs the quote form's product picker
// and customer selector. Each list response is passed through as raw JSON
// (its exact field shape isn't committed to server-side); see
// service/catalog.go and flowpos/client.go.
type CatalogHandler struct {
	catalog *service.CatalogService
}

func NewCatalogHandler(catalog *service.CatalogService) *CatalogHandler {
	return &CatalogHandler{catalog: catalog}
}

func (h *CatalogHandler) Products(c *gin.Context) {
	data, err := h.catalog.ListProducts(c.Request.Context(), tenantID(c), c.Query("search"))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"products": data})
}

func (h *CatalogHandler) Product(c *gin.Context) {
	data, err := h.catalog.GetProduct(c.Request.Context(), tenantID(c), c.Param("slug"))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"product": data})
}

func (h *CatalogHandler) Categories(c *gin.Context) {
	data, err := h.catalog.ListCategories(c.Request.Context(), tenantID(c), c.Query("search"))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"categories": data})
}

func (h *CatalogHandler) Customers(c *gin.Context) {
	data, err := h.catalog.ListCustomers(c.Request.Context(), tenantID(c), c.Query("search"))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"customers": data})
}

func (h *CatalogHandler) CreateCustomer(c *gin.Context) {
	var in struct {
		Name  string  `json:"name" binding:"required"`
		Email string  `json:"email" binding:"required"`
		Phone *string `json:"phone"`
	}
	if err := c.ShouldBindJSON(&in); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	data, err := h.catalog.CreateCustomer(c.Request.Context(), tenantID(c), in.Name, in.Email, in.Phone)
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusCreated, gin.H{"customer": data})
}

func (h *CatalogHandler) Locations(c *gin.Context) {
	data, err := h.catalog.ListLocations(c.Request.Context(), tenantID(c), c.Query("search"))
	if err != nil {
		fail(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"locations": data})
}
